torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,272 @@
+
+  bool is_scalar() const {
+    return type_ == TensorDataContainerType::Scalar;
+  }
+
+  const c10::Scalar& scalar() const {
+    TORCH_CHECK(
+      is_scalar(),
+      "Can only call `scalar()` on a TensorDataContainer that has `is_scalar() == true`");
+    return scalar_;
+  }
+
+  bool is_init_list() const {
+    return type_ == TensorDataContainerType::InitList;
+  }
+
+  const std::initializer_list<TensorDataContainer>& init_list() const {
+    TORCH_CHECK(
+      is_init_list(),
+      "Can only call `init_list()` on a TensorDataContainer that has `is_init_list() == true`");
+    return init_list_;
+  }
+
+  bool is_tensor() const {
+    return type_ == TensorDataContainerType::Tensor;
+  }
+
+  const at::Tensor& tensor() const {
+    TORCH_CHECK(
+      is_tensor(),
+      "Can only call `tensor()` on a TensorDataContainer that has `is_tensor() == true`");
+    return tensor_;
+  }
+
+  const std::vector<int64_t>& sizes() const {
+    return sizes_;
+  }
+
+  const c10::ScalarType& scalar_type() const {
+    return scalar_type_;
+  }
+
+  at::Tensor convert_to_tensor(const at::TensorOptions& options) const {
+    if (is_scalar()) {
+      at::AutoNonVariableTypeMode non_var_type_mode(true);
+      return at::scalar_tensor(scalar_, options.is_variable(false));
+    } else if (is_init_list()) {
+      // NOTE: Here we explicitly choose to initialize the tensor on CPU first,
+      // fill each element of the tensor, and then move the tensor to the desired
+      // device. For CUDA device, this approach only involves 1 CUDA kernel launch,
+      // and is much faster than initializing the tensor on CUDA first and then
+      // filling each element of it (which involves `N` CUDA kernel launches where
+      // `N` is the number of the elements in the tensor).
+      at::Tensor tensor = ([&]() {
+        at::AutoNonVariableTypeMode non_var_type_mode(true);
+        return at::empty(sizes_, options.device(at::kCPU).is_variable(false));
+      })();
+      fill_tensor(tensor);
+      return tensor.to(options.device());
+    } else if (is_tensor()) {
+      return tensor_.to(options);
+    } else {
+      TORCH_INTERNAL_ASSERT(false, "Invalid TensorDataContainer type");
+    }
+  }
+
+  void pretty_print_recursive(std::ostream& stream) const {
+    if (is_scalar()) {
+      // NOTE: There is no `operator<<` overload for `at::kBFloat16` type,
+      // and we need to convert it to `float` type using `operator float()` function
+      // defined in c10/util/BFloat16.h.
+      if (scalar_type_ == at::ScalarType::BFloat16) {
+        stream << static_cast<float>(scalar_.to<c10::BFloat16>());
+      } else {
+        AT_DISPATCH_ALL_TYPES_AND2(at::kBool, at::kHalf, scalar_type_, "TensorDataContainer_pretty_print_scalar", [&] {
+          stream << scalar_.to<scalar_t>();
+        });
+      }
+    } else if (is_init_list()) {
+      stream << "{";
+      for (const TensorDataContainer* it = init_list_.begin(); it != init_list_.end(); it++) {
```

Comment:
nit: you could `const auto` here but this is fine

Reply:
oh, I see. "auto" or "const auto*" should work but it's fine to keep it as `const TensorDataContainer*`